src/modules/rlm_redis/cluster.c

```diff
@@ -0,0 +1,2254 @@
+			return CLUSTER_OP_IGNORED;
+		}
+		fr_strerror_printf("Unknown client error");
+		return CLUSTER_OP_FAILED;
+
+	case REDIS_RCODE_SUCCESS:
+		break;
+	}
+
+	if (reply->type != REDIS_REPLY_ARRAY) {
+		fr_strerror_printf("Bad response to \"cluster slots\" command, expected array got %s",
+				   fr_int2str(redis_reply_types, reply->type, "<UNKNOWN>"));
+		return CLUSTER_OP_BAD_INPUT;
+	}
+
+	/*
+	 *	Clustering configured but no slots set
+	 */
+	if (reply->elements == 0) {
+		fr_strerror_printf("Empty response to \"cluster slots\" command (zero length array)");
+		return CLUSTER_OP_BAD_INPUT;
+	}
+
+	/*
+	 *	Validate the complete map set before returning.
+	 */
+	for (i = 0; i < reply->elements; i++) {
+		size_t		j;
+		redisReply	*map;
+
+		map = reply->element[i];
+		if (map->type != REDIS_REPLY_ARRAY) {
+			fr_strerror_printf("Cluster map %zu is wrong type, expected array got %s",
+				   	   i, fr_int2str(redis_reply_types, map->type, "<UNKNOWN>"));
+		error:
+			fr_redis_reply_free(reply);
+			return CLUSTER_OP_BAD_INPUT;
+		}
+
+		if (map->elements < 3) {
+			fr_strerror_printf("Cluster map %zu has too few elements, expected at least 3, got %zu",
+					   i, map->elements);
+			goto error;
+		}
+
+		/*
+		 *	Key slot start
+		 */
+		if (map->element[0]->type != REDIS_REPLY_INTEGER) {
+			fr_strerror_printf("Cluster map %zu key slot start is wrong type, expected integer got %s",
+					   i, fr_int2str(redis_reply_types, map->element[0]->type, "<UNKNOWN>"));
+			goto error;
+		}
+
+		if (map->element[0]->integer < 0) {
+			fr_strerror_printf("Cluster map %zu key slot start is too low, expected >= 0 got %lli",
+					   i, map->element[0]->integer);
+			goto error;
+		}
+
+		if (map->element[0]->integer > KEY_SLOTS) {
+			fr_strerror_printf("Cluster map %zu key slot start is too high, expected <= "
+					   STRINGIFY(KEY_SLOTS) " got %lli", i, map->element[0]->integer);
+			goto error;
+		}
+
+		/*
+		 *	Key slot end
+		 */
+		if (map->element[1]->type != REDIS_REPLY_INTEGER) {
+			fr_strerror_printf("Cluster map %zu key slot end is wrong type, expected integer got %s",
+					   i, fr_int2str(redis_reply_types, map->element[1]->type, "<UNKNOWN>"));
+			goto error;
+		}
+
+		if (map->element[1]->integer < 0) {
+			fr_strerror_printf("Cluster map %zu key slot end is too low, expected >= 0 got %lli",
+					   i, map->element[1]->integer);
+			goto error;
+		}
+
+		if (map->element[1]->integer > KEY_SLOTS) {
+			fr_strerror_printf("Cluster map %zu key slot end is too high, expected <= "
+					   STRINGIFY(KEY_SLOTS) " got %lli", i, map->element[1]->integer);
+			goto error;
+		}
+
+		if (map->element[1]->integer < map->element[0]->integer) {
+			fr_strerror_printf("Cluster map %zu key slot start/end out of order.  "
+					   "Start was %lli, end was %lli", i, map->element[0]->integer,
+					   map->element[1]->integer);
+			goto error;
+		}
+
+		/*
+		 *	Master node
+		 */
+		if (map->element[2]->type != REDIS_REPLY_ARRAY) {
```

Comment:
Probably should remove if statement as it looks like master node never gets validated.